### PR TITLE
LFVM: Test integer overflow from stack values

### PR DESCRIPTION
**Pull request by LuisPH3**
part of #751 

This test checks operations that fetch values (256 Bits) from the stack for 64 bits operators. 
Both CALL and CREATE family of operators are too complicated for this batch testing, and require to be tested on its own. 

Included refactor:

- Reorder some operations code to prioritize check of operators overflow before more specific operator checks. 
- Remove hard-coded maxuint64 uses. 
- Remove ignored err values, return them instead.

### CODE REVIEW

File: go/interpreter/lfvm/instructions_test.go [hunk 1]
```diff
@@ -1624,3 +1624,189 @@ func TestGeneralCall_ResultIsWrittenToStack(t *testing.T) {
...
+			inputs: []inputData{
+				inputStack(maxUint256, zero, zero),
+				inputStack(zero, maxUint256, zero),
+				inputStack(one, zero, maxUint256),
+			},
+		},
+		CODECOPY: {
+			implementation: opCodeCopy,
+			inputs: []inputData{
+				inputStack(maxUint256, zero, zero),
```
Comment (HerbertJordan) [inline go/interpreter/lfvm/instructions_test.go:1691]: This only checks the `size` for an overflow, but not the two other parameters, which are offsets.

In the code, size overflows seem to be treated differently than offset overflows. Size overflows lead to `errOverflow` while other overflows seem to result in out-of-gas errors. I am wondering whether not all overflows related to offsets in memory accessed would lead to out-of-gas errors, and thus should not be explicitly checked.

tl;dr: are some of those extra checks for overflow errors that would actually also be called by out-of-gas errors?

File: go/interpreter/lfvm/instructions_test.go [hunk 2]
```diff
@@ -1624,3 +1624,189 @@ func TestGeneralCall_ResultIsWrittenToStack(t *testing.T) {
...
+					if want, got := errOverflow, err; want != got {
+						t.Fatalf("unexpected error, wanted %v, got %v", want, got)
+					}
+				}
+			})
+		})
+	}
+}
+
+func TestInstructions_CopyOpRoundStackValuesOnOverflow(t *testing.T) {
```
Comment (HerbertJordan) [inline go/interpreter/lfvm/instructions_test.go:1750]: If I understand correctly, this tests make sure that overflows in offsets are ignored if the size is zero. If so, this should be reflected in the name. I do not think it has anything to do with rounding ...

**Review by HerbertJordan — CHANGES_REQUESTED** [review]
I like the code cleanup, bit I am not sure the property that is tested by the new test is an actual requirement or just an implementation artifact.

In all cases I reviewed in this PR, it seems like numeric overflows are related to offsets in the memory, in code, or in various data buffers (call return data or some code). If there is such an overflow, setting the 64-bit value derived from the 256-bit value on the stack to the maximum 64-bit value (saturation semantic) should automatically lead to an actual error (e.g. out-of-gas when attempting to access memory) or to a correct handling (returning zero values when reading from a buffer, since no buffer is long enough).

Thus, the dedicated handling of overflows and reporting those as errors seems to be redundant. It actually introduces the risk of identifying an overflow as an issue that should actually be ignored if, for instance, the offset of a memory access overflows but the size of the memory offset is zero. Such a case should not be marked as an error.

If you agree, I would suggest to update the tests to check for any kind of errors and also use 2^64 as a test input, since this value would be reduced to 0 if the overflow check is incorrectly implemented, although it should be considered equivalent to 2^64-1 = MaxUint64.

**Comment by HerbertJordan** [comment]
Another way to think about it: if we would update the member functions of the [memory](https://github.com/Fantom-foundation/Tosca/blob/main/go/interpreter/lfvm/memory.go) implementation to deal with `uint256.Int` instead of `uint64` many of the checks in the op-codes dealing with the memory would move into the memory implementation. And there, "overflows" would be naturally caught by out-of-gas errors.

File: go/interpreter/lfvm/instructions_test.go [hunk 3]
```diff
@@ -1624,3 +1624,189 @@ func TestGeneralCall_ResultIsWrittenToStack(t *testing.T) {
...
+				inputStack(one, one, maxUint256),
+			},
+		},
+		RETURNDATACOPY: {
+			implementation: opReturnDataCopy,
+			inputs: []inputData{
+				inputStack(maxUint256, zero, zero),
+				inputStack(zero, maxUint256, zero),
+				inputStack(one, zero, maxUint256),
+			},
```
[the remark this answers is not in the document]
Comment (LuisPH3) [reply go/interpreter/lfvm/instructions_test.go:1686]: I find it quite cumbersome to write arrays of arrays and the code begins to look like Perl, I will revert this code as it seems to be that it is not idiomatic.

File: go/interpreter/lfvm/instructions_test.go [hunk 4]
```diff
@@ -1624,3 +1624,189 @@ func TestGeneralCall_ResultIsWrittenToStack(t *testing.T) {
...
+		JUMP: {
+			implementation: opJump,
+			inputs: []inputData{
+				inputStack(maxUint256),
+			},
+		},
+		JUMPI: {
+			implementation: opJumpi,
+			inputs: []inputData{
+				inputStack(one, maxUint256),
```
[the remark this answers is not in the document]
Comment (LuisPH3) [reply go/interpreter/lfvm/instructions_test.go:1646]: There should be a second test for the jump Ops that can ignore the data.

File: go/interpreter/lfvm/instructions_test.go [hunk 5]
```diff
@@ -1624,3 +1624,189 @@ func TestGeneralCall_ResultIsWrittenToStack(t *testing.T) {
...
+			},
+		},
+		MSTORE: {
+			implementation: opMstore,
+			inputs: []inputData{
+				inputStack(zero, maxUint256),
```
[the remark this answers is not in the document]
Comment (LuisPH3) [reply go/interpreter/lfvm/instructions_test.go:1652]: after rebase, i can start using some tool Facu wrote.

**Comment by LuisPH3** [comment]
This PR is closed because revision exposed that the code has structural flaws that should be refactored insted of tested.
